src/pyodide/internal/serializeJsModule.ts

```diff
@@ -0,0 +1,105 @@
+/* eslint-disable prefer-rest-params */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-argument */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import { IS_CREATING_SNAPSHOT } from 'pyodide-internal:metadata';
+
+export type SerializedJsModule = {
+  jsModule: true;
+  moduleName: string;
+  accessorList: string[];
+};
+
+const importName = Symbol('importName');
+const getAccessorList = Symbol('getAccessorList');
+const getObject = Symbol('getObject');
+const getPrototypeOfKey = 'Reflect.getProtoTypeOf';
+
+export function maybeSerializeJsModule(
+  obj_: any,
+  modules: Set<string>
+): SerializedJsModule | undefined {
+  const obj = obj_ as
+    | { [importName]: string; [getAccessorList]: string[] }
+    | undefined;
+  const moduleName = obj?.[importName];
+  if (!moduleName) {
+    return undefined;
+  }
+  modules.add(moduleName);
```

Comment:
It is slightly uncomfortable to me that this function has a side effect of updating the `modules` set, which needs to be drilled down to this function from a very top-level `makeLinearMemorySnapshot`. But I guess it is trivial thing.

Reply:
Yeah it's unfortunate but the dynamic import is async and deserialization has to be async. Perhaps we should change it upstream so the deserialization can be async, that would make use cases like this simpler.